src/transaction/declare_v2.rs

```diff
@@ -320,6 +320,13 @@ impl DeclareV2 {
             })
             .map_err(|e| TransactionError::SierraCompileError(e.to_string()))?;
 
+        let casm_class_hash = compute_casm_class_hash(casm_class)?;
+        if casm_class_hash != self.compiled_class_hash {
```

Comment:
Does it make sense to always have this check? Maybe we could have it as an optional _feature_, or add it [only for debugging](https://doc.rust-lang.org/reference/conditional-compilation.html#debug_assertions).

Reply:
No, it's not necessary to _always_ check this. We want to check this whenever a `casm contract class` and `casm class hash` is sent from the outside, because otherwise we couldn't guarantee that the compiled contract is really the corresponded to that sierra contract class. It wouldn't be a problem if the contract was already compiled on our side. I'll work on that next, but preferred to have this implemented first.

Reply:
I'm in favor of making it a debug assertion. Either we computed it and they are the same by construction, or it was passed via one of the unchecked functions and the responsibility lies in the caller. The point of that was not having to compute the hash in the first place.
Unless I'm misunderstanding it, of course.

Reply:
Another option may be to don't compile and check at all, and take the casm contract class and class hash as valid. But sound pretty insecure. We should think a better way of guarantee that.

Reply:
It is insecure, but that's the compromise made for performance and that's why we have to be extra explicit and opt-in. If you hold it wrong, it breaks.
The only way to guarantee correctness is to compute stuff ourselves, but then we only moved the costs and increased complexity for no gain for the user. If we're gonna check, just compute it on construction and never again.

Reply:
Think of it like using `unsafe` in Rust. The keyword doesn't mean you can do whatever you like, but that you go to battle without the shield, it's your job to dodge the blows now. When the programmer opts for passing the hash themselves, it's their job to make sure it's the correct one, and if they get hurt they'll deal with it.